File: Sources/KiboCore/Conversion/RunSplitter.swift
import Foundation

/// A maximal stretch of one kind of text. Mixed mode judges and converts one run at a time, so
/// where the boundaries fall decides what gets judged together.
struct Run: Equatable, Sendable {
    enum Script: Sendable {
        /// Thai script.
        case thai
        /// Printable ASCII — letters, digits, and punctuation alike.
        case latin
        /// Everything else. Never converted, never judged, passed through exactly.
        case neutral
    }

    let script: Script
    let text: String

    init(script: Script, text: String) {
        self.script = script
        self.text = text
    }
}

/// Splits text into maximal same-script runs.
enum RunSplitter {

    /// ASCII punctuation stays *inside* the Latin run rather than being treated as neutral, and
    /// that is deliberate: `;` is the `ว` key, so pulling it out would break `l;ylfu` → `สวัสดี`.
    /// The cost is that a run like `2024` is nominally "Latin"; `LatinOrthography` is what keeps
    /// it from being converted.
    ///
    /// Space is excluded from the Latin range on purpose. It makes whitespace a run boundary, so
    /// two words either side of a space are judged separately instead of as one blob.
    private static func script(of scalar: UnicodeScalar) -> Run.Script {
        switch scalar.value {
        case 0x0E00...0x0E7F: return .thai
        case 0x21...0x7E: return .latin
        default: return .neutral
        }
    }

    /// Runs always rejoin to the input exactly — nothing is dropped, reordered, or normalised.
    static func split(_ input: String) -> [Run] {
        var runs: [Run] = []
        var current = String.UnicodeScalarView()
        var currentScript: Run.Script?

        for scalar in input.unicodeScalars {
            let scalarScript = script(of: scalar)
            if scalarScript != currentScript, currentScript != nil {
                runs.append(Run(script: currentScript!, text: String(current)))
                current = String.UnicodeScalarView()
            }
            currentScript = scalarScript
            current.append(scalar)
        }
        if let currentScript {
            runs.append(Run(script: currentScript, text: String(current)))
        }
        return runs
    }
}

File: Windows/Kibo.Core/Conversion/RunSplitter.cs
using System.Text;

namespace Kibo.Core;

/// <summary>Splits text into maximal same-script runs. The port of <c>RunSplitter.swift</c>.</summary>
internal static class RunSplitter
{
    /// <summary>
    /// ASCII punctuation stays <i>inside</i> the Latin run rather than being treated as neutral,
    /// and that is deliberate: <c>;</c> is the <c>ว</c> key, so pulling it out would break
    /// <c>l;ylfu</c> → <c>สวัสดี</c>. The cost is that a run like <c>2024</c> is nominally "Latin";
    /// <see cref="LatinOrthography"/> is what keeps it from being converted.
    /// <para>
    /// Space is excluded from the Latin range on purpose. It makes whitespace a run boundary, so
    /// two words either side of a space are judged separately instead of as one blob.
    /// </para>
    /// <para>
    /// Curly quotes and long dashes count as Latin even though they sit outside the ASCII range,
    /// because an OS put them there in place of <c>'</c>, <c>"</c> and <c>-</c>. Left neutral,
    /// they cut a word in two — <c>don’t</c> split into three runs — and <see cref="RunJudge"/>
    /// never saw a word to judge.
    /// </para>
    /// </summary>
    private static Script ScriptOf(Rune scalar) => scalar.Value switch
    {
        >= 0x0E00 and <= 0x0E7F => Script.Thai,
        >= 0x21 and <= 0x7E => Script.Latin,
        _ => TypographicSubstitutes.Contains(scalar) ? Script.Latin : Script.Neutral,
    };

    /// <summary>
    /// Runs always rejoin to the input exactly — nothing is dropped, reordered, or normalised.
    /// </summary>
    public static IReadOnlyList<Run> Split(string input)
    {
        var runs = new List<Run>();
        var current = new StringBuilder();
        Script? currentScript = null;

        foreach (var scalar in input.EnumerateRunes())
        {
            var scalarScript = ScriptOf(scalar);
            if (currentScript is { } open && scalarScript != open)
            {
                runs.Add(new Run(open, current.ToString()));
                current.Clear();
            }
            currentScript = scalarScript;
            current.Append(scalar.ToString());
        }
        if (currentScript is { } last)
        {
            runs.Add(new Run(last, current.ToString()));
        }
        return runs;
    }
}
